src/gui_common/ModalManager.cs

```diff
@@ -85,7 +85,13 @@ public void MakeModal(TopLevelContainer popup)
         if (modalStack.Contains(popup))
             return;
 
-        originalParents[popup] = popup.GetParent();
+        var parent = popup.GetParent();
+        originalParents[popup] = parent;
+
+        // Listen for when the parent is removed from the tree so the modal can be removed as well.
+        parent.Connect(Node.SignalName.TreeExiting, Callable.From(() => OnParentLost(popup)),
```

Comment:
I think it would be worth it to track what parents are connected in this way? So that they can be unconnected on normal close (and not require leaving the tree). Or at least made so that no duplicate connections are made by this method.

Reply:
Good catch. I've added the code to track and remove signals when the modal is demoted.